modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MetricRegistry.java

```diff
@@ -0,0 +1,300 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.metrics;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import org.apache.ignite.internal.util.FilteringIterator;
+import org.apache.ignite.internal.util.TransformingIterator;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Metric registry. Metrics source (see {@link MetricSource} must be registered in this metrics registry after initialization
+ * of corresponding component and must be unregistered in case of component is destroyed or stopped. Metrics registry also
+ * provides access to all enabled metrics through corresponding metrics sets. Metrics registry lifetime is equal to the node lifetime.
+ * <br>
+ * Implements an {@link Iterable} over the metric sets for enabled metric sources.
+ */
+public class MetricRegistry implements Iterable<MetricSet> {
+    private final Lock lock = new ReentrantLock();
+
+    /** Map of metric sources' names to tuples of registered sources with metric sets, if enabled. */
+    private volatile Map<String, IgniteBiTuple<MetricSource, MetricSet>> sources = new TreeMap<>();
+
+    /** Version always should be changed on metrics enabled/disabled action. */
+    private volatile long version;
+
+    /**
+     * Register metric source. It must be registered in this metrics registry after initialization of corresponding component
+     * and must be unregistered in case of component is destroyed or stopped, see {@link #unregisterSource(MetricSource)}.
+     * By registering, the metric source isn't enabled implicitly.
+     *
+     * @param src Metric source.
+     * @throws IllegalStateException If metric source with the given name already exists.
+     */
+    public void registerSource(MetricSource src) {
+        modifySources(sources -> {
+            IgniteBiTuple<MetricSource, MetricSet> s = new IgniteBiTuple<>(src, null);
+
+            IgniteBiTuple<MetricSource, MetricSet> old = sources.putIfAbsent(src.name(), s);
+
+            if (old != null) {
+                throw new IllegalStateException("Metrics source with given name already exists: " + src.name());
+            }
+
+            // Now we sure that this metric source wasn't registered before.
+            assert !src.enabled() : "Metric source shouldn't be enabled before registration in registry.";
```

Comment:
Why should not metric source enabled before registering? It is too strict limitation because the `MetricSource`'s contract doesn't prohibit this.

Reply:
Once again, `MetricSource`'s contract doesn't protect a developer from this error. Moreover, metrics source could be disabled by configuration but it is possible to register not enabled metrics source.